### PR TITLE
Convert types if possible on WRITE

### PR DESCRIPTION
**Pull request by kasper-piskorski**
## What is the goal of this PR?

Allow for implicit type conversion if possible when doing WRITEs. As a result queries like:

```
define resource sub attribute, datatype double;
insert $x 10 isa attribute;
```

Will succeed as there exists a unambiguous conversion. Previously we didn't allow that.

### CODE REVIEW

File: server/src/server/kb/concept/AttributeImpl.java [hunk 1]
```diff
@@ -52,11 +52,13 @@ private AttributeImpl(VertexElement vertexElement, AttributeType<D> type, D valu
     }
 
     public static <D> AttributeImpl<D> create(VertexElement vertexElement, AttributeType<D> type, D value) {
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/AttributeImpl.java]: How can `value` ever have a different `Class` than was is defined in the schema when `type` is defined to be `AttributeType<D>`? This method signature already bounds the type, no? @kasper-piskorski @marco-scoppetta

File: server/src/server/kb/concept/TypeConverter.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    public static class DateConverter extends TypeConverter<LocalDateTime> {
+        @Override
+        public Object serialise(LocalDateTime value) {
+            return value.atZone(ZoneId.of("Z")).toInstant().toEpochMilli();
+        }
+
+        @Override
+        public LocalDateTime deserialise(Object value) {
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/TypeConverter.java]: Would it be better to restrict the input argument to always be `Long`?

File: server/src/server/kb/concept/TypeConverter.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            if (value instanceof LocalDateTime) return (LocalDateTime) value;
+            if (value instanceof Long) return LocalDateTime.ofInstant(Instant.ofEpochMilli((Long) value), ZoneId.of("Z"));
+            throw new ClassCastException();
+        }
+    }
+
+    public static class DoubleConverter extends TypeConverter<Double> {
+        @Override
+        public Double deserialise(Object value) {
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/TypeConverter.java]: I think the input arguments here should definitely be restricted to `Number`? Along with every other `TypeConverter`, for that matter.

File: server/src/server/kb/concept/TypeConverter.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            if (value instanceof Number){
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/TypeConverter.java]: what happens when you cast a `Long` to a `Double` or `Float`?

File: test-integration/graql/query/BUILD [hunk 5]
```diff
@@ -20,161 +20,159 @@ load("@graknlabs_build_tools//checkstyle:rules.bzl", "checkstyle_test")
...
 )
 
 java_test(
     name = "graql-delete-it",
-    test_class = "grakn.core.graql.query.GraqlDeleteIT",
+    size = "medium",
     srcs = ["GraqlDeleteIT.java"],
+    test_class = "grakn.core.graql.query.GraqlDeleteIT",
     deps = [
-        "//common:common",
```
Comment (haikalpribadi) [inline test-integration/graql/query/BUILD]: you should put back the target names, @kasper-piskorski . As you can see in other target definitions, we're making it a convention to write the full `//package/name:target-name`

File: server/src/server/kb/concept/TypeConverter.java (hunk 3 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/TypeConverter.java]: We still want to throw if it is an unsupported conversion/not a Number, so I don't think so. Also that way the whole conversion logic is in one place.

File: test-integration/graql/query/BUILD (hunk 5 above)
Comment (kasper-piskorski) [reply test-integration/graql/query/BUILD]: it's not me, it's bazel/intellij

File: server/src/server/kb/concept/TypeConverter.java (hunk 4 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/TypeConverter.java]: I don't get the question. If it's a Number (so Long as well) it gets converted to a Double.

File: server/src/server/kb/concept/TypeConverter.java (hunk 2 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/TypeConverter.java]: Same argument as for Double. If we constrained it to Long we would need to convert somewhere else/make sure it's a Long. Here we treat general objects so the whole conversion logic sits in a single place.

File: server/src/server/kb/concept/AttributeImpl.java (hunk 1 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/AttributeImpl.java]: D is `Object` at run-time as generics in java only work at compile-time, there's no way to enforce the specific type at run-time.

File: server/src/server/kb/concept/TypeConverter.java (hunk 3 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/TypeConverter.java]: Ok, i changed my mind!

File: server/src/server/kb/concept/ValueConverter.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        public TARGET convert(SOURCE value) { return (TARGET) value;}
+    }
+
+    public static class DateConverter extends ValueConverter<Object, LocalDateTime> {
+
+        @Override
+        public LocalDateTime convert(Object value) {
+            if (value instanceof LocalDateTime) return (LocalDateTime) value;
+            else if (value instanceof Long) {
+                LocalDateTime.ofInstant(Instant.ofEpochMilli((Long) value), ZoneId.of("Z"));
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/ValueConverter.java]: Shouldn't there be a `return` statement here, @kasper-piskorski ?

File: server/src/server/kb/concept/Serialiser.java [hunk 7]
```diff
@@ -78,26 +72,23 @@ public VALUE serialise(VALUE value) {
...
 
-    public static class Date extends Serialiser<LocalDateTime, Long> {
-
-        Date() {
-            super();
-        }
+    public static class Date extends Serialiser<java.time.LocalDateTime, Long> {
 
         @Override
         public java.lang.Long serialise(LocalDateTime value) {
```
Comment (haikalpribadi) [inline server/src/server/kb/concept/Serialiser.java]: this seems to be remnant from old code, but might as well remove `java.lang.` prefix too.

File: server/src/server/kb/concept/AttributeImpl.java (hunk 1 above)
Comment (haikalpribadi) [reply server/src/server/kb/concept/AttributeImpl.java]: So if `type` has class `AttributeType<Long>`, the last argument `value` is not bound to `Long` as well?

File: server/src/server/kb/concept/ValueConverter.java (hunk 6 above)
Comment (kasper-piskorski) [reply server/src/server/kb/concept/ValueConverter.java]: good spot!

**Pull request merged**